### PR TITLE
Complete refactor of the timeCacher implementation

### PR DESCRIPTION
**Pull request by iulianpascalau**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- complete refactor of the timeCacher implementation
- removed duplicated stubs
  
## Proposed Changes
- Implemented a new variant of the timeCacher that do not rely on the timeCache implementation but a timeCacheCore that contains the common code between the timeCacher and timeCache implementations. This will assure that we no longer have deadlock conditions due to the usage of more than one mutex. Also, the code should be easier to be revised, debugged and extended.

## Testing procedure
- standard system test, pending customized test in a different setting

### CODE REVIEW

File: storage/timecache/timeCacher_test.go [hunk 1]
```diff
@@ -278,15 +276,74 @@ func TestMapTimeCacher_UnRegisterHandler(t *testing.T) {
...
+			case 10:
+				_ = tc.MaxSize()
+			case 11:
+				tc.RegisterHandler(nil, "")
+			case 12:
+				tc.UnRegisterHandler("")
+			case 13:
+				_ = tc.Close()
+			default:
+				assert.Fail(t, "test setup error, change this line 'switch idx%6{'")
```
Comment (sstanculeanu) [inline storage/timecache/timeCacher_test.go:333]:
```suggestion
				assert.Fail(t, "test setup error, change this line 'switch idx%14{'")
```

File: storage/timecache/timeCacher.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+const minDuration = time.Second
+
+// ArgTimeCacher is the argument used to create a new timeCacher instance
+type ArgTimeCacher struct {
+	DefaultSpan time.Duration
+	CacheExpiry time.Duration
+}
+
+// timeCacher implements a map cache with eviction and inner TimeCacher
```
Comment (sstanculeanu) [inline storage/timecache/timeCacher.go:22]: perhaps
```suggestion
// timeCacher implements a time cacher with automatic sweeping mechanism
```

File: storage/timecache/timeCacheCore_test.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+			case 1:
+				tcc.sweep()
+			case 2:
+				_ = tcc.has(fmt.Sprintf("key%d", idx))
+			case 3:
+				_ = tcc.len()
+			case 4:
+				tcc.clear()
+			default:
+				assert.Fail(t, "test setup error, change this line 'switch idx%6{'")
```
Comment (sstanculeanu) [inline storage/timecache/timeCacheCore_test.go:36]:
```suggestion
				assert.Fail(t, "test setup error, change this line 'switch idx%5{'")
```

File: storage/timecache/timeCacher_test.go (hunk 1 above)
Comment (iulianpascalau) [reply storage/timecache/timeCacher_test.go:333]: fixed

File: storage/timecache/timeCacher.go (hunk 2 above)
Comment (iulianpascalau) [reply storage/timecache/timeCacher.go:22]: done

**Review by sstanculeanu — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by dragos-rebegea — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Pull request merged**